bpf/proxy/syncer.go

```diff
@@ -664,31 +665,70 @@ func getSvcNATKey(svc k8sp.ServicePort) (nat.FrontendKey, error) {
 	}
 
 	key := nat.NewNATKey(ip, uint16(port), proto)
-
 	return key, nil
 }
 
-func (s *Syncer) writeSvc(svc k8sp.ServicePort, svcID uint32, count, local int) error {
-	key, err := getSvcNATKey(svc)
+func getSvcNATKeyLBSrcRange(svc k8sp.ServicePort) ([]nat.FrontendKey, error) {
+	var keys []nat.FrontendKey
+	ipaddr := svc.ClusterIP()
+	port := svc.Port()
+	loadBalancerSourceRanges := svc.LoadBalancerSourceRanges()
+	log.Debugf("loadbalancer %v", loadBalancerSourceRanges)
+	proto, err := protoV1ToInt(svc.Protocol())
 	if err != nil {
-		return err
+		return keys, err
+	}
+	key := nat.NewNATKey(ipaddr, uint16(port), proto)
+	keys = append(keys, key)
+	for _, src := range loadBalancerSourceRanges {
+		key := nat.NewNATKeySrc(ipaddr, uint16(port), proto, ip.MustParseCIDROrIP(src).(ip.V4CIDR))
+		keys = append(keys, key)
 	}
+	return keys, nil
+}
 
+func (s *Syncer) writeSvcNatKey(key nat.FrontendKey, val nat.FrontendValue) error {
+	log.Debugf("bpf map writing %s:%s", key, val)
+	return s.bpfSvcs.Update(key[:], val[:])
+}
+
+func (s *Syncer) writeSvc(svc k8sp.ServicePort, svcID uint32, count, local int) error {
 	affinityTimeo := uint32(0)
 	if svc.SessionAffinityType() == v1.ServiceAffinityClientIP {
 		affinityTimeo = uint32(svc.StickyMaxAgeSeconds())
 	}
-
+	var affkey nat.FrontEndAffinityKey
 	val := nat.NewNATValue(svcID, uint32(count), uint32(local), affinityTimeo)
-
-	log.Debugf("bpf map writing %s:%s", key, val)
-	if err := s.bpfSvcs.Update(key[:], val[:]); err != nil {
-		return errors.Errorf("bpfSvcs.Update: %s", err)
+	if len(svc.LoadBalancerSourceRanges()) == 0 {
+		key, err := getSvcNATKey(svc)
+		if err != nil {
+			return err
+		}
+		err = s.writeSvcNatKey(key, val)
+		if err != nil {
+			return errors.Errorf("bpfSvcs.Update: %s", err)
+		}
+		copy(affkey[:], key[4:12])
+	} else {
+		keys, err := getSvcNATKeyLBSrcRange(svc)
+		if err != nil {
+			return err
+		}
+		for index, key := range keys {
+			if index == 0 {
+				err = s.writeSvcNatKey(key, nat.NewNATValue(math.MaxUint32, uint32(0), uint32(local), affinityTimeo))
+				copy(affkey[:], key[4:12])
+			} else {
+				err = s.writeSvcNatKey(key, val)
+			}
+			if err != nil {
+				return errors.Errorf("bpfSvcs.Update: %s", err)
+			}
+		}
 	}
```

Comment:
How are updates handled?  If the set of allowed IPs changes or is removed completely?

Reply:
Have UTs for addition, updation, deletion of source IP ranges.

Reply:
Indeed, looks like the cases I was worried about are handled by the surrounding code.  It must convert an update into a delete then an add?

Reply:
After all new services are added, we delete the ones that no longer exist, however, the LB takes a different path. When the source IPs range changes, the service itself changes and we always first delete a service in this case.

The LB handling is done in a slightly different way then the other services, which may or may not be OK - normally, the ClusterIP service is the "primary service" and then other services are treated as "derived". This implementation diverts from that pattern, but given the different purpose of the clusterIP (a black hole) and derived services (the actual NAT) in this case it may be ok.

Note that the "derived" lpm entries never make it in the `newSvcMap/prevSvcMap`

One thing I am not sure will work well is if we have an `ExternalIP` assigned. The iptables kube-proxy also does the filtering on the IP. I think the way we have it know, we would bypass the filtering as the ExternalIP would be just derived from the ClusterIP service. We should also cover this in a unittest.

Regarding recovery after crash, we need to cover that in a UT (another reason why not to reuse an existing service - tuning it into an LB - in the test). Since the keys will have the the src IP prefix part, we will know that it is an LB source IP range match entry, but it does not seem we handle that case and we need to make sure that we do not delete these entries because we cannot match them with the new set of services. And we must make sure we delete stale if the LB does not exist anymore.